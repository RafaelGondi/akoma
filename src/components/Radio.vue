<template>
	<span class="cds-radio">
		<!--
			Evento emitido quando o Radio muda seu estado.
			@event change
			@type {Event}
		-->
		<input
			type="radio"
			:id="value"
			:value="value"
			:disabled="disabled"
			:name="name"
			:checked="isChecked"
			@change="$emit('change', $event.target.value)"
		>
		<label
			class="cds-radio__content"
			:for="value"
		>
			<!-- @slot Slot padrão para renderização de conteúdo customizado da label -->
			<slot>
				<span
					class="cds-radio__label"
				>
					{{ label }}
				</span>
			</slot>
		</label>
	</span>
</template>

<script>
export default {
	model: {
		prop: 'modelValue',
		event: 'change',
	},

	props: {
		/**
		 * A prop usada como v-model para monitorar a seleção do Radio
		*/
		modelValue: {
			default: null,
			required: true,
		},
		/**
		 * A prop usada como valor associado ao Radio. O que será atribuído
		 * v-model quando essa opção for selecionado
		*/
		value: {
			default: null,
			required: true,
		},
		/**
		 * A prop usada para controle do input pelo name. Definir mesmo name para um grupo de opções.
		*/
		name: {
			type: String,
			default: '',
		},
		/**
		 * Define a label do input, o conteúdo que é exibido para descrever o Radio.
		 */
		label: {
			type: String,
			default: '',
		},
		/**
		 * Controla o status do radio
		 */
		disabled: {
			type: Boolean,
			default: false,
			required: false,
		},
		/**
		 * Estado do radio em relação a validação
		 */
		invalid: {
			type: Boolean,
			default: false,
			required: false,
		},
	},

	computed: {
		isChecked() {
			return this.modelValue == this.value;
		},
	},
};
</script>

<style lang="scss">
@import '../assets/sass/app.scss';

.cds-radio [type="radio"]:checked,
.cds-radio [type="radio"]:not(:checked) {
	position: absolute;
	left: -9999px;
}

.cds-radio [type="radio"]:checked + label,
.cds-radio [type="radio"]:not(:checked) + label
{
	position: relative;
	padding: pl(7);
	cursor: pointer;
	line-height: 20px;
	display: inline-block;
	color: $n-800;
}

.cds-radio [type="radio"]:checked + label:before,
.cds-radio [type="radio"]:not(:checked) + label:before {
	content: '';
	position: absolute;
	left: 0;
	top: 0;
	width: 20px;
	height: 20px;
	border: 1px solid $n-200;
	border-radius: 100%;
	background: $n-0;
}

.cds-radio [type="radio"]:checked + label:after,
.cds-radio [type="radio"]:not(:checked) + label:after {
	content: '';
	width: 12px;
	height: 12px;
	background: $gp-500;
    border: 1px solid $gp-500;
	position: absolute;
	top: 4px;
	left: 4px;
	border-radius: 100%;
	-webkit-transition: all 0.3s ease;
	transition: all 0.3s ease;
}

.cds-radio [type="radio"]:not(:checked) + label:after {
	opacity: 0;
	-webkit-transform: scale(0);
	transform: scale(0);
}

.cds-radio [type="radio"]:checked + label:after {
	opacity: 1;
	-webkit-transform: scale(1);
	transform: scale(1);
}

.cds-radio input[type="radio"][disabled="disabled"] + label:before {
	-webkit-transition: all 0.3s ease;
	transition: all 0.3s ease;
    background: $n-30;
    color: $n-30;
    cursor: not-allowed;
}

.cds-radio input[type="radio"][invalid="invalid"] + label:before {
	border: 2px solid $rc-400;
}
</style>
