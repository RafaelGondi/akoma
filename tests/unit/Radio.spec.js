import { mount, createLocalVue } from '@vue/test-utils';
import Radio from '../../src/components/Radio.vue';
import flushPromises from 'flush-promises';

import BootstrapVue from 'bootstrap-vue';
const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	const wrapper = mount(Radio, {
		localVue,
		propsData: {
			modelValue: null,
			value: 'option-1',
			label: 'Opção 1',
		},
	});

	expect(wrapper).toMatchSnapshot();
});

test('if updates v-model with value of radio on click', async () => {
	let modelValue = null;
	const wrapper = mount(Radio, {
		localVue,
		propsData: {
			modelValue,
			value: 'option-1',
			label: 'Opção 1',
			disabled: false,
			invalid: false,
		},
	});

	wrapper.find('.cds-radio__content').trigger('click');

	await flushPromises();

	expect(wrapper.emitted().change).toBeTruthy();
});
