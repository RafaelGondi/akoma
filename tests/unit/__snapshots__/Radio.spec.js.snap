// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<span class="cds-radio"><input type="radio" id="option-1" name="" value="option-1"> <label for="option-1" class="cds-radio__content"><span class="cds-radio__label">
				Opção 1
			</span></label></span>
`;
